addons/mass_mailing/controllers/main.py

```diff
@@ -55,3 +87,51 @@ def full_url_redirect(self, code, stat_id, **post):
 
         request.env['link.tracker.click'].add_click(code, request.httprequest.remote_addr, country_code, stat_id=stat_id)
         return werkzeug.utils.redirect(request.env['link.tracker'].get_url_from_code(code), 301)
+
+    @http.route('/mail/mailing/blacklist/check', type='json', auth='none')
+    def check_blacklist(self, email):
+        if email:
+            record = request.env['mail.blacklist'].sudo().with_context(active_test=False).search([('email', '=ilike', email)])
+            if record['active']:
+                return 'found'
+            return 'not found'
+        return 'error'
+
+    @http.route('/mail/mailing/blacklist/add', type='json', auth='none')
+    def add_to_blacklist(self, mailing_id, res_id, email, token):
+        mailing = request.env['mail.mass_mailing'].sudo().browse(mailing_id)
+        if mailing._unsubscribe_token(res_id, email) != token:
+            return 'unauthorized'
+        if email:
+            recipient_name = 'lead' if mailing.mailing_model_real == 'crm.lead' \
+                else 'partner' if mailing.mailing_model_real == 'res.partner' \
+                else 'recipient'
+            request.env['mail.blacklist'].sudo()._add(email, """The %s asked 
```

Comment:
Here probably just use _message_log after updating the record, as we are about to remove the reason field.